fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java

```diff
@@ -605,14 +607,84 @@ public Set<String> getUpdatedPartitionNamesOfOlapTable(OlapTable baseTable) {
         return result;
     }
 
-    private boolean isLessThanMVRewriteStaleness() {
+    /**
+     * @return Return max timestamp of all table's max refresh timestamp
+     *          which is computed by checking all its partitions' modified time.
+     */
+    public Optional<Long> maxBaseTableRefreshTimestamp() {
+        long maxRefreshTimestamp = -1;
+        for (BaseTableInfo baseTableInfo : baseTableInfos) {
+            Table baseTable = baseTableInfo.getTable();
+            if (baseTable instanceof MaterializedView) {
+                MaterializedView mv = (MaterializedView) baseTable;
+                if (!mv.isStalenessSatisfied()) {
+                    return Optional.empty();
+                }
+                Optional<Long> maxPartitionRefreshTimestamp =
+                        mv.getPartitions().stream().map(Partition::getVisibleVersionTime).max(Long::compareTo);
+                if (!maxPartitionRefreshTimestamp.isPresent()) {
+                    return Optional.empty();
+                }
+                maxRefreshTimestamp = Math.max(maxPartitionRefreshTimestamp.get(), maxRefreshTimestamp);
+            } else if (baseTable instanceof OlapTable) {
+                OlapTable olapTable = (OlapTable) baseTable;
+                Optional<Long> maxPartitionRefreshTimestamp =
+                        olapTable.getPartitions().stream().map(Partition::getVisibleVersionTime).max(Long::compareTo);
+                if (!maxPartitionRefreshTimestamp.isPresent()) {
+                    return Optional.empty();
+                }
+                maxRefreshTimestamp = Math.max(maxPartitionRefreshTimestamp.get(), maxRefreshTimestamp);
+            } else if (baseTable instanceof HiveTable) {
+                HiveTable hiveTable = (HiveTable) baseTable;
+                Map<String, com.starrocks.connector.PartitionInfo> partitionNameWithPartition =
+                        PartitionUtil.getPartitionNameWithPartitionInfo(hiveTable);
+                Optional<Long> maxPartitionRefreshTimestamp =
+                        partitionNameWithPartition.values().stream().map(com.starrocks.connector.PartitionInfo::getModifiedTime)
+                                .max(Long::compareTo);
+                if (!maxPartitionRefreshTimestamp.isPresent()) {
+                    return Optional.empty();
+                }
+                maxRefreshTimestamp = Math.max(maxPartitionRefreshTimestamp.get(), maxRefreshTimestamp);
+            } else if (baseTable instanceof View) {
+                // continue
+            } else {
```

Comment:
what if iceberg and hudi table?

Reply:
Good question. 
To be more safe, now We don't support it for staleness. We can support it later?